crypto/evp/digest.c

```diff
@@ -140,7 +140,22 @@ int EVP_DigestInit_ex(EVP_MD_CTX *ctx, const EVP_MD *type, ENGINE *impl)
     }
     if (ctx->flags & EVP_MD_CTX_FLAG_NO_INIT)
         return 1;
-    return ctx->digest->init(ctx);
+    if (ctx->digest->init(ctx) == 0)
+        return 0;
+
+    if (ctx->pctx) {
+        int r;
+        r = EVP_PKEY_CTX_ctrl(ctx->pctx, -1, EVP_PKEY_OP_TYPE_SIG,
+                              EVP_PKEY_CTRL_DIGESTSETUP, 0, ctx);
+
+        if (r == -2) {
+            ERR_clear_error();
```

Comment:
The problem is `EVP_PKEY_CTX_ctrl` does this:

```
    if (ret == -2)
        EVPerr(EVP_F_EVP_PKEY_CTX_CTRL, EVP_R_COMMAND_NOT_SUPPORTED);

    return ret;
```

And here in digest.c we are swallowing the error return, since only SM2 uses this DIGESTSETUP PKEY flag. If we do not also clear the error state, it causes code downstream to fail. Specifically I noticed failures in the TLS layer tests unless I cleared the error state here, presumably because somewhere TLS looks at the error stack in order decide if something failed.

Reply:
Perhaps, `ERR_set_mark()`, `ERR_pop_to_mark()`, `ERR_clear_last_mark()` would be better?

Reply:
Seems so since that avoids wiping out the entire error stack. Thanks, didn't know about these.